mpc85xx: add support for Kernel 6.12

CHKDSK88 opened the pull request:
This PR is based on #18778 and I use almost every commits from there.

@blocktrron prepared everything, but I prefer to use `./scripts/kernel_bump.sh` for bump preparation.

This Pull-Request adds support for Kernel 6.12 for Freescale mpc85xx QorIQ based boards.

Compile-Tested:

    p1010
    p1020
    p2020

Run-Tested: p2020-rdb (p2020), br200-wp (p1010)

blocktrron committed:
mpc85xx: introduce 6.12 testing kernel

Signed-off-by: David Bauer <mail@david-bauer.net>

blocktrron committed:
mpc85xx: refresh kernel 6.12 patches

Signed-off-by: David Bauer <mail@david-bauer.net>

blocktrron committed:
mpc85xx: use strscpy instead of strlcpy

Usage of strlcpy results in a compilation error. strscpy is preferred
anyways and equivalent for our case.

Signed-off-by: David Bauer <mail@david-bauer.net>

blocktrron committed:
mpc85xx: refresh kernel config

Signed-off-by: David Bauer <mail@david-bauer.net>

CHKDSK88 committed:
kernel/mpc85xx: Restore kernel files for v6.6

This is an automatically generated commit which aids following Kernel patch
history, as git will see the move and copy as a rename thus defeating the
purpose.

For the original discussion see:
https://lists.openwrt.org/pipermail/openwrt-devel/2023-October/041673.html

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 committed:
kernel/mpc85xx: Create kernel files for v6.12 (from v6.6)

This is an automatically generated commit.

When doing `git bisect`, consider `git bisect --skip`.

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

Shine- commented:
Tested with Watchguard T10 (with #16776 applied, otherwise T10 is completely broken).
Apparently the FSL_IFC driver is missing again (already happened in 6.6), so it can't find its NAND device and therefore can't mount its root fs.

Shine- commented:
Only searched around in Google so far, and all I could find is what's already present (ie. explicitly setting `CONFIG_MEMORY=y` and `CONFIG_FSL_IFC=y`). I'll continue searching and testing during the week. That I'm rather illiterate when it comes to kernel config and all its back-and-forth-and-cross dependencies, won't make it easy for me, though.

Shine- commented:
So far, I could find out that the driver is there:
<details><summary>/sys/bus/platform/drivers/</summary>

```
root@OpenWrt:~# ls -l /sys/bus/platform/drivers/
drwxr-xr-x    2 root     root             0 Nov  4 22:05 alarmtimer
drwxr-xr-x    2 root     root             0 Nov  4 22:05 ehci-platform
drwxr-xr-x    2 root     root             0 Nov  4 22:05 fsl,ifc-nand
drwxr-xr-x    2 root     root             0 Nov  4 22:05 fsl-ehci
drwxr-xr-x    2 root     root             0 Nov  4 22:05 fsl-gianfar
drwxr-xr-x    2 root     root             0 Nov  4 22:05 fsl-ifc
drwxr-xr-x    2 root     root             0 Nov  4 22:05 fsl-lbc
drwxr-xr-x    2 root     root             0 Nov  4 22:05 fsl-pci
drwxr-xr-x    2 root     root             0 Nov  4 22:05 fsl-pq_mdio
drwxr-xr-x    2 root     root             0 Nov  4 22:05 fsl-usb2-mph-dr
drwxr-xr-x    2 root     root             0 Nov  4 22:05 fsl_espi
drwxr-xr-x    2 root     root             0 Nov  4 22:05 gpio-export
drwxr-xr-x    2 root     root             0 Nov  4 22:05 gpio-keys
drwxr-xr-x    2 root     root             0 Nov  4 22:05 gpio-mpc8xxx
drwxr-xr-x    2 root     root             0 Nov  4 22:05 leds-gpio
drwxr-xr-x    2 root     root             0 Nov  4 22:05 mpc-i2c
drwxr-xr-x    2 root     root             0 Nov  4 22:05 mpc85xx_l2_err
drwxr-xr-x    2 root     root             0 Nov  4 22:05 mpc85xx_mc_err
drwxr-xr-x    2 root     root             0 Nov  4 22:05 mpc85xx_pci_err
drwxr-xr-x    2 root     root             0 Nov  4 22:05 of_serial
drwxr-xr-x    2 root     root             0 Nov  4 22:05 physmap-flash
drwxr-xr-x    2 root     root             0 Nov  4 22:05 qe-ic
drwxr-xr-x    2 root     root             0 Nov  4 22:05 reg-dummy
drwxr-xr-x    2 root     root             0 Nov  4 22:05 serial8250
drwxr-xr-x    2 root     root             0 Nov  4 22:05 simple-pm-bus
drwxr-xr-x    2 root     root             0 Nov  4 22:05 spi_gpio
drwxr-xr-x    2 root     root             0 Nov  4 22:05 talitos
drwxr-xr-x    2 root     root             0 Nov  4 22:05 ucc_geth
root@OpenWrt:~# ls -l /sys/bus/platform/drivers/fsl-ifc/
--w-------    1 root     root          4096 Nov  4 22:05 bind
--w-------    1 root     root          4096 Nov  4 22:05 uevent
--w-------    1 root     root          4096 Nov  4 22:05 unbind

```
</details>

but the actual device node is missing in 6.12, while in 6.6, it's present:

<details><summary>6.6: "ffe1e000.ifc" is present</summary>

```
root@OpenWrt:~# ls -l /sys/devices/platform/
drwxr-xr-x    3 root     root             0 Nov  4 22:50 Fixed MDIO bus.0
drwxr-xr-x    4 root     root             0 Nov  4 22:50 buttons
drwxr-xr-x    5 root     root             0 Nov  4 22:50 ffe0a000.pcie
drwxr-xr-x    3 root     root             0 Nov  4 22:50 ffe1e000.ifc
drwxr-xr-x    9 root     root             0 Nov  4 22:50 leds
drwxr-xr-x    3 root     root             0 Nov  4 22:50 reg-dummy
drwxr-xr-x    2 root     root             0 Nov  4 22:50 regulatory.0
drwxr-xr-x    4 root     root             0 Nov  4 22:50 serial8250
drwxr-xr-x    2 root     root             0 Nov  4 22:50 serial8250.0
drwxr-xr-x   28 root     root             0 Nov  4 22:50 soc@ffe00000
-rw-r--r--    1 root     root          4096 Nov  4 22:50 uevent
```
</details>
<details><summary>6.12: "ffe1e000.ifc" is missing</summary>

```
root@OpenWrt:~# ls -l /sys/devices/platform/
drwxr-xr-x    3 root     root             0 Nov  4 21:49 Fixed MDIO bus.0
drwxr-xr-x    4 root     root             0 Nov  4 21:49 buttons
drwxr-xr-x    5 root     root             0 Nov  4 21:49 ffe0a000.pcie
drwxr-xr-x    9 root     root             0 Nov  4 21:49 leds
drwxr-xr-x    3 root     root             0 Nov  4 21:49 reg-dummy
drwxr-xr-x    2 root     root             0 Nov  4 22:05 regulatory.0
drwxr-xr-x    3 root     root             0 Nov  4 21:49 serial8250
drwxr-xr-x    2 root     root             0 Nov  4 21:49 serial8250.0
drwxr-xr-x   29 root     root             0 Nov  4 21:49 soc@ffe00000
-rw-r--r--    1 root     root          4096 Nov  4 22:05 uevent
```
</details>

No idea how to proceed from here, but I assume one could enable the `ifc: ifc@ffe1e000` node also in the DTS of non-NAND equipped devices, and it should show up in `/sys/devices/platform/` despite no NAND flash device present?

Please note that all my tests are done with (an updated version of) #16776 applied, not with vanilla OpenWrt. Otherwise the T10 wouldn't boot at all, even in 6.6.

neheb commented:
may or may not have found the issue: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.12.57&id=acb354fe97e5aa6d9534b601ce18ef7866f25c4d

Shine- commented:
[Current documentation](https://www.kernel.org/doc/Documentation/devicetree/bindings/memory-controllers/fsl/fsl%2Cifc.yaml) has an example with `ifc` as child of `soc`. To be able to use it that way would mean a major rewrite of the affected DTS, as far as I understand (which is "only" Watchguard T10 and Sophos Red 15w at this time, as far as I could see).

neheb commented:
Best of luck. To be clear, I don't own hardware of this target.

neheb commented:
Anyway, I submitted https://lore.kernel.org/linuxppc-dev/20251105205524.17362-1-rosenp@gmail.com/T/#u .

edit: the commit author's email bounced. I'm guessing he's not at NXP anymore.

CHKDSK88 committed:
mpc85xx: kernel: 6.12: fix issue with ifc nodes in P1010

In upstream [0] ifc node compatible was changed and it broke ifc at all.
Rosen Penev send revert commit to upstream [1] which repair it.

This patch add this pending patch to mpc85xx.

[0] https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=0bf51cc9e9e57a751b4c5dacbfa499ba5cd8bd72
[1] https://patchwork.ozlabs.org/project/linuxppc-dev/patch/20251105205524.17362-1-rosenp@gmail.com/

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

Shine- commented:
Soooo....

Since I could obtain a 2-pack of Sophos RED 15w for dirt cheap, I took the liberty to test with that device as well.

Sophos RED 15w is broken since 24.10.0 and needs the following DTS patch to boot any non-initramfs image:
<details><summary>Sophos RED 15w partition boundary fix</summary>

Explanation: Newer versions of `mtd` don't allow exceeding partition boundaries anymore. This already broke several Linksys IPQxxxx devices' installation instructions. Here, it unfortunately breaks the whole image. Therefore, move the offending partitions one level up, so they won't run over their parent partition's end anymore.

In case any one of you feels like it, this patch would need to be committed into 24.10 and Main to fix booting for this device.
```diff
diff --git a/target/linux/mpc85xx/files/arch/powerpc/boot/dts/red-15w-rev1.dts b/target/linux/mpc85xx/files/arch/powerpc/boot/dts/red-15w-rev1.dts
index 19bb8023ed..166aada816 100644
--- a/target/linux/mpc85xx/files/arch/powerpc/boot/dts/red-15w-rev1.dts
+++ b/target/linux/mpc85xx/files/arch/powerpc/boot/dts/red-15w-rev1.dts
@@ -168,20 +168,16 @@
 				oem-partition@300000 {
 					reg = <0x300000 0x1900000>;
 					label = "sophos-os1";
+				};
 
-					compatible = "fixed-partitions";
-					#address-cells = <1>;
-					#size-cells = <1>;
-
-					partition@0 {
-						reg = <0x0 0x800000>;
-						label = "kernel";
-					};
+				partition@300000 {
+					reg = <0x300000 0x800000>;
+					label = "kernel";
+				};
 
-					partition@800000 {
-						reg = <0x800000 0x7500000>;
-						label = "ubi";
-					};
+				partition@b00000 {
+					reg = <0xb00000 0x7500000>;
+					label = "ubi";
 				};
 
 				oem-partition@1c00000 {
```
</details>

In addition, also already since 24.10 (kernel 6.6), the FIT load address has to be adjusted due to memory overlapping of compressed/uncompressed kernel image. I have fixed the Wiki page, so this is already accounted for.

Other than that, kernel 6.12 boots fine, ethernet works, routing works, WiFi works. LGTM.

neheb commented:
sounds like more devices need a fix like that.

edit: nvm.

Shine- commented:
> sounds like more devices need a fix like that.

Like what? Kernel uncompressed size is an issue with the whole target, agreed. Funny how kernel 6.6 was introduced, apparently without any actual testing, that broke like all devices in this target ;-). And nobody even noticed, figure how many poeple are actually using this target (read: 0).

MTD boundaries I really don't understand (doesn't belong here, though). I tried to figure out how people could honestly provide installation instructions that, judging from the timestamp they were provided, couldn't possibly work, not in Stable nor in Main...

Shine- commented:
Huh?

This is a great, underestimated platform. A Single-Core 533MHz P1010 CPU kills e.g. any 800MHz MIPS32 CPU. And then, there are the multi-core versions, too, like P1020, and the 64bit flavors like T10xx (not supported yet).
Now imagine if the crypto engine were actually supported...

Shine- commented:
Why would the MTD fix not be a proper solution?
- it keeps the MTD numbers as-is, so updating from previous versions, as well as any installation instructions, can remain
- it keeps the OEM MTD partitions available, so they can be used to back up and restore the OEM firmware

Of course, all OEM partition definitions could be removed completely, they're not used at all.

The boot loader does not look for anything, the original installation instructions (provided by @blocktrron) load the image from NAND with fixed offsets, and so does my update. I personally don't like the original instructions of messing with the stock uboot-env, but that's how this device's support was introduced, and how it should remain in order to keep it compatible for existing users.

My above patch does the **minimum possible** in order to make the device bootable with kernel 6.6 and 6.12 - so **I** can boot kernel 6.12 to test this very PR. I tried retaining the original committer's decisions as much as possible. I'm not trying to suggest anything new, and I'm **especially not** trying to persuade anybody of doing anything like I want it to be done (I don't care either way!). So if you don't like it, feel very free to PR and commit your own version of the fix - not my business. I just bought these devices so I can help @CHKDSK88 test his kernel 6.12 PR :-)

Edit: Just FYI: Reducing UBI size to 0x1100000 breaks booting for existing users, because it will then refuse mounting.

neheb commented:
That makes no sense. The original size is wrong and either should be reduced to 0x1100000 or the extra partitions should be removed and os1 expanded.

I don't know as is if the UBI volume itself is set to 0x1100000 despite the bigger size.

Shine- commented:
UBI partition size is properly set to 0x7500000 - ie. from where the root fs partiton starts to the end of the flash.
Again, **do whatever you want**. Not my business. My patch could boot 6.6 and 6.12, and I could back up and restore the stock partitions that way. Have a better idea? Do it! I don't care at all! Again, the **only** reason I bought these devices was to test this PR. And my goal was to keep the OEM partitions able to back up - as obviously intended by @blocktrron - and to restore them as well (so I can resell these things anytime).

Shine- commented:
> @Shine- I assume that gpio-button change would break others users :.

Proof?

neheb commented:
I mean you added

+			linux,input-type = <EV_SW>;
+			debounce-interval = <60>;

only to the wdr4900. I assume the others would need similar treatment.

Shine- commented:
I was purely catering to #13552, and didn''t even own any other `mpc85xx` device at that time. All that mattered to me was that it was working. Whether the `debounce-interval = <60>;` is even needed or could be ommitted, no idea, I didn't try. It worked, I kept it. The OP of #13552 didn't report back, so I didn't see any reason to re-test. That's how it is :-)

If anybody wants to actually commit that, let me know, I can easily re-test with and without `debounce-interval`.

neheb commented:
@CHKDSK88 I just noticed while looking at dts.

br200-wp is listed as p1010 but the dts file has fsl/p1020si-pre.dtsi and fsl/p1020si-post.dtsi

CHKDSK88 commented:
> BR200 is a P1011, ie. a P1020 with one core disabled. So, the DTSI is correct.

Exactly. P1011 is P1020 castrated to one core. P1010 is much different.

@neheb 

Could you ping upstream about revert patch? T4240 is also affected.
